libnavigation-base/src/main/java/com/mapbox/navigation/base/internal/route/NavigationRouteEx.kt

```diff
@@ -194,17 +192,22 @@ private fun List<LegStep>.updateSteps(
     return result
 }
 
-private fun DirectionsResponse.Builder.updateWaypoints(
+private fun buildNewWaypoints(
     oldWaypoints: List<DirectionsWaypoint>?,
     updatedWaypoints: List<DirectionsWaypoint?>?,
-): DirectionsResponse.Builder {
+): List<DirectionsWaypoint>? {
     if (oldWaypoints == null || updatedWaypoints == null) {
-        return this
+        return oldWaypoints
     }
     val newWaypoints = oldWaypoints.mapIndexed { index, oldWaypoint ->
-        updatedWaypoints.getOrNull(index) ?: oldWaypoint
```

Comment:
Why are we opting to update only unrecognized fields instead of taking the whole updated waypoint? If in the future more fields were updated in waypoints we'd have to copy each of them manually as well.

Reply:
As @VysotskiVadim pointed out, if we updated the whole waypoint, we introduce a technical possibility that locations would change or some other stuff that we don't expect to change.
As in legs we have a white-list of what's being updated (annotations, incidents and closures) and what we are ready for to be updated, I agree that here we should have white-list as well.

Reply:
> As in legs we have a white-list of what's being updated (annotations, incidents and closures) and what we are ready for to be updated, I agree that here we should have white-list as well.

We only have a white-list for these elements because the refresh response does not contain the full leg, so we need to independently update each relevant element.

This is not the case for waypoints. The waypoint structure and data in the refresh response is exactly the same as the waypoiont structure in the original route response (according to the private docs) so we can just swap them out. If there were suddenly different coordinates returned, that would be a bug on the Nav API side.

Reply:
Alright, makes sense.